feat: 메서드 콜그래프 저장 및 조회 & 1차 고도화

GunsanHaribo opened the pull request:
## PR 타입
- [x] 메서드 메타 데이터 추출
- [x] 메서드 콜 그래프 생성 내부 정책 보강
- [x] 멀티모듈 구성 변경

---

## 개요
- 바이트 코드에서 메서드 메타 데이터 추출
- 메서드 콜 그래프 생성

---

## 변경 사항
- 바이트 코드 기반으로 메서드 메타 데이터 저장
- 메서드의 Caller, Callee를 통해서 메서드 콜그래프 저장(간선테이블을 사용해 저장)
- 연결시 내부 정책은 링크 참조(링크 추가예정)
- git clone에 사용된 폴더는 자원으로 처리해서 에러나 완료시 자동 삭제되도록 AutoClosable 도입

---

## 주의 사항
- 모든 자바 레포에서 작동하지 않을 수도 있음, 현재 OpensourceReader에서만 작동

## 잔여 버그
- [ ]  바이트코드 파싱안되는 버그 수정 필요
- [ ]  자기자신 필드로가질 수 있는 엔티티들은 무조건 자기자신 못넣게 해야함
- [ ]  멀티모듈일떄 패키지랑 클래스이름 중복되면 어떻게 할건지 생각하기
- [ ]  클래스안에 Private 클래스가 있을떄는 Content에 타입으로 띄워지는지 확인해야함
    - [ ]  private한 클래스들이나 내부 클래스들은 어떻게 들어오는지 체크 필요



Close #14

GunsanHaribo committed:
feat: 파일 저장시, 메서드 정보 추출 후 따로 저장

GunsanHaribo committed:
feat: Java 소스 코드에서 메서드 호출 대상(path + 메서드명) 추출 기능 추가

GunsanHaribo committed:
feat: CodeMethodCallEdge 생성

GunsanHaribo committed:
feat: MethodFlowService createFlow 기능 추가

GunsanHaribo committed:
feat: CodeMethodMetaData 메서드 파라미터 추출 기능 도메인 정적생성팩터리메서드에 추가

GunsanHaribo committed:
feat: ExpressionTypeResolver 일부 구현

GunsanHaribo committed:
feat: 현재 메서드에서 사용하는 다른 메서드(outgoing)메서드 저장

GunsanHaribo committed:
refactor: 소스코드 구조(AST활용)해서 Call graph를 만드는 파일 삭제

소스코드에서 바이트코드기반 로직으로 수정을 해서 삭제합니다.

GunsanHaribo committed:
feat: 현재 메서드를 사용하는 다른 메서드들(ingoing) 을 연결합니다

GunsanHaribo committed:
refactor: 테스트코드 수정

- OpenSourceRepoFacadeTest @Disabled 위치변경
- CodeMethodMetaDataTest 결과 수정

GunsanHaribo committed:
refactor: 사용하지 않는 레포지토리 클래스 삭제 및 수정

GunsanHaribo committed:
refactor: bare-repo 워킹트리 저장 Path 고정

GunsanHaribo committed:
refactor: 스프링 어노케이션 정리 및 인터페이스 추가

- 불필요한 어노테이션 정리
- 맥락상 서비스가 아닌 컴포넌트로 수정

GunsanHaribo committed:
refactor: path 만드는 방법 통일

- ".java"를 일일이 합쳤던 로직을 DTO에 위임

GunsanHaribo committed:
refactor: 메서드 시그니처 만드는 방식 통일

GunsanHaribo committed:
refactor: createMethodCallGraph 정리

GunsanHaribo committed:
refactor: FileUtil 패키지이동

GunsanHaribo committed:
refactor: OpenSourceRepoFacade, OpenSourceRepoService 정리

GunsanHaribo committed:
refactor: OpenSourceRepoContentMethodCallGraphService 수정

- OpenSourceRepoClassMethodService 인터페이스 생성 및 로직 수정
- 메서드 콜 수집로직 추출

GunsanHaribo committed:
refactor: OpenSourceRepoContentName 레코드로 변경

GunsanHaribo committed:
feat: code_method, method_call_edge flyway용 SQL추가 및 수정

GunsanHaribo committed:
feat: String 확장자 -> Extension 객체로 수정

GunsanHaribo committed:
feat: .java 붙이는 로직 통일

GunsanHaribo committed:
refactor: MethodCall DTO 정리 및 테스트 추가

GunsanHaribo committed:
test: CodeMethod 테스트 추가

GunsanHaribo committed:
refactor: 서비스레이어 테스트 코드 윤곽 작성

GunsanHaribo committed:
test: OpenSourceMethodExtractor 테스트 코드 작성

GunsanHaribo committed:
refactor: ClassMethodCallResult -> MethodCallsOfClass DTO 이름 수정

GunsanHaribo committed:
refactor: CallGraphAnalyzerTest 추가 및 코드수정

GunsanHaribo committed:
refactor: 바이트 코드 추출 패키지 이동 및 OpenSourceRepoClassMethodService 테스트 윤곽 작성

GunsanHaribo committed:
refactor: LocalOpenSourceRepoClassMethodService 일부수정

GunsanHaribo committed:
docs: CallGraph 정책 V1 작성

GunsanHaribo committed:
refactor: 소스코드내의 메서드 구조 추출시, 소스코드에서 바이트코드 기반 추출로 변경

GunsanHaribo committed:
refactor: OpenSourceContentFactory로 리펙터링

GunsanHaribo committed:
test: OpenSourceCodeMethodExtractorTest 보완

GunsanHaribo committed:
refactor: CallGraphAnalyzer -> ClassStructureExtractor 메서드 이름 수정 및 테스트 코드 보강

GunsanHaribo committed:
refactor: MethodCallInfo DTO 테스트 삭제

GunsanHaribo committed:
refactor: 패키지 이름 수정

GunsanHaribo committed:
refactor: open_source_repo_content에 className 추가

GunsanHaribo committed:
refactor: createMethodCallGraph 수정

- 외부라이브러리 호출을 버리지않고, external로 저장

GunsanHaribo committed:
refactor: 패키지 구조 및 클래스 이름 수정

GunsanHaribo committed:
feat: codeMethod callgraph 조회기능 추가

GunsanHaribo committed:
feat: codeMethod callgraph API 조회기능 추가

GunsanHaribo committed:
refactor: 레포생성 -> 코드구조 생성 -> 코드 메서드 그래프 생성 최종 테스트 및 수정

GunsanHaribo committed:
refactor: 패키지 구조 수정

GunsanHaribo committed:
refactor: Content, Method 생성 로직 수정

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/dto/OpenSourceFileInfo.java:
```diff
@@ -0,0 +1,3 @@
+package com.opensourcereader.core.analysis.dto;
+
+public record OpenSourceFileInfo(String path, String typeNumber, String rawText) {}
```
이거 typeNumber면 Interger타입 아닌가하는데
String타입으로 하신 이유가 있으실까요?

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/dto/callgraph/ClassInfo.java:
```diff
@@ -0,0 +1,31 @@
+      String signature,
+      String superName,
+      String[] interfaces) {
+    return new ClassInfo(
```
getInterfaces 메서드를 삭제하고 합치는 건 어떻습니까?
테스트 목적을 위해서 분리시킨거라면 괜찮지만 만약 그것이 아니라면 하나로 합쳐도 괜찮을 듯 합니다.
가령
```java

  public static ClassInfo of(
      int version,
      int access,
      String className,
      String signature,
      String superName,
      String[] interfaces) {
    List<String> interfaceNames = (interfaces == null) 
        ? List.of() 
        : Arrays.asList(interfaces);
    return new ClassInfo(
        version, access, className, signature, superName, interfaceNames);
  }
```
코드를 읽는 사람 입장에서는 of를 보다가 갑자기 getInterfaces로 시선을 돌려야하는데 이게 특별한 로직도 아니고 단지 null체크인데 굳이 시선을 분산시킬 필요가 없어보인다고 생각듭니다.

또 마틴 파울러의 리펙토링 책, 6.2 함수 인라인하기를 보시면
간접 호출을 과하게 하는 것을 피하라고 되어있습니다.
이를 토대로 null체크 메서드를 없애고 삼항연산자를 넣는게 좋지않을까 합니다.

아래는 한국인이 쓴 리뷰글입니다. 6.2 함수 인라인하기를 보시면 좋겠습니다.
https://github.com/Youngerjesus/refactoring-2nd-edition/tree/master/document/chapter-06

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/config/JavaParserConfig.java:
```diff
@@ -0,0 +1,19 @@
+  @Bean
+  public JavaParser javaParser() {
+    ParserConfiguration parserConfiguration =
+        new ParserConfiguration().setLanguageLevel(LanguageLevel.JAVA_17);
```
제가 알기로는 자바 파서 역할이
1. 오픈소스 레포를 가져온다.
2. 레포를 분석한다.
3. 컴파일한다.
4. 메서드를 자바파서로 분석한다.

이런 걸로 알고있습니다.
현재 자바 17로 되어있는데 이게 자바 17버전 밑이면 괜찮은데
만약 자바 21이거나 그 이상이 되면 문법적으로 괜찮은데 오류가 터질수도 있을거 같습니다.
해당 파트에 대해서 한 번 생각해보시면 좋을거 같네요.
혹은 
```java
        new ParserConfiguration().setLanguageLevel(LanguageLevel.BLEEDING_EDGE);
```
이렇게 BLEEDING_EDGE를 통해 자바 파서가 현재 지원하는 최신 문법으로 파서하는 것도 괜찮은 방법같습니다.

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/dto/callgraph/CodeMethodExtractResult.java:
```diff
@@ -0,0 +1,44 @@
+    Integer endLine) {
+
+  public static CodeMethodExtractResult of(
+      DeclaredMethodInfo declaredMethodInfo, SourceCodeParseResult codeParseResult) {
```
간접 함수를 사용하지 않고

```java
public record CodeMethodExtractResult(
    String methodName,
    AccessModifier modifier,
    EnumSet<NonAccessModifier> nonAccessModifiers,
    String returnType,
    List<String> paramTypes,
    Integer startLine,
    Integer endLine) {

  public static CodeMethodExtractResult of(
      DeclaredMethodInfo declaredMethodInfo, SourceCodeParseResult codeParseResult) {
    Integer startLine = null;
    Integer endLine = null;
    if (codeParseResult != null){
      startLine = codeParseResult.startLine();
      endLine = codeParseResult.endLine();
    }
    
    return new CodeMethodExtractResult(
        declaredMethodInfo.methodName(),
        declaredMethodInfo.accessModifier(),
        declaredMethodInfo.nonAccessModifiers(),
        declaredMethodInfo.methodDescriptor().methodReturnType(),
        declaredMethodInfo.methodDescriptor().argumentTypes(),
        startLine,
        endLine);
  }
}

```

이렇게 메서드 호출없이 한 번에 처리할 수 있을거 같습니다.
출처 :
https://github.com/Youngerjesus/refactoring-2nd-edition/tree/master/document/chapter-06

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/dto/callgraph/method/DeclaredMethodInfo.java:
```diff
@@ -0,0 +1,42 @@
+      String descriptor,
+      String genericSignature,
+      String[] exceptions) {
+    return new DeclaredMethodInfo(
```
CodeMethodExtractResult와 동일합니다.
간접함수 날리고 삼항연산자로 처리하면 좋을거같습니다.

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/entity/codemethod/MethodCallOrigin.java:
```diff
@@ -0,0 +1,6 @@
+package com.opensourcereader.core.analysis.entity.codemethod;
+
+public enum MethodCallOrigin {
```
MethodCallOrigin  이거랑 MethodOrigin 이거 안에 값은 같은데 무엇때문에 이렇게 두 enum으로 분리하셨나요?

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/infra/bytecode/ClassStructureVisitor.java:
```diff
@@ -0,0 +1,77 @@
+      String[] interfaces) {
+    this.classInfo =
+        new ClassInfo(
+            version, access, className, signature, superName, Arrays.stream(interfaces).toList());
```
어... ClassInfo 클래스에서는 Null 체크를 때린걸로 기억하는데
여기서는 Null체크를 안하시네요....
문제 없으면 넘어가도 무방한데
한 번 확인해주세요

BrotherMountain commented on core/src/main/java/com/opensourcereader/core/analysis/service/impl/CodeMethodGraphQueryServiceImpl.java:
```diff
@@ -0,0 +1,31 @@
+    CodeMethod codeMethod =
+        codeMethodRepository
+            .findWithOutgoingGraphById(codeMethodId)
+            .orElseThrow(IllegalArgumentException::new);
```
이거 오류 클래스 만들지 않고 기본 오류 클래스 쓰실건가요?
의도한 바였다면 괜찮고 혹시 깜빡하셨다면 추가하는 것도 괜찮아보입니다.

BrotherMountain commented on core/core-domain/src/main/java/com/opensourcereader/core/analysis/infra/parser/SourceCodeParser.java:
```diff
@@ -0,0 +1,100 @@
+
+@Component
+@RequiredArgsConstructor
+public class SourceCodeParser {
```
개인적인 생각입니다.
이게 나중에 자바파서에 심각한 문제가 생기거나 업데이트가 중단되거나 하는 일이 생기면 
만약 해당 라이브러리에 의존적이라면 뜯어고쳐야합니다.
해서 인터페이스와 구현체로 구현해서 쉽게 갈아끼울 수 있게 만들면 어떨까합니다.

BrotherMountain reviewed (commented):
@GunsanHaribo 
ClassStructureVisitor 쪽 확인만 해주시면 Approve하겠습니다.

GunsanHaribo committed:
chore: 코드 리펙토링 윤곽 일부 작성

GunsanHaribo committed:
docs: 콜그래프 정책 보완

GunsanHaribo committed:
refactor: RepoContent에 DeclaredType 타입 추가

GunsanHaribo committed:
docs: 콜그래프 문서 보완

GunsanHaribo committed:
refactor: CodeMethodCallGraphService 뼈대 수정

GunsanHaribo committed:
feat: MethodPolymorphicDispatchService 다형성 인터페이스 연결기능 추가

GunsanHaribo committed:
feat: MethodPolymorphicDispatchService 다형성 상속 연결기능 추가

GunsanHaribo committed:
refactor: MethodCallLinkService 로직 수정

GunsanHaribo committed:
test: OpenSourceContentFactory 테스트 작성

GunsanHaribo committed:
refactor: DeclaredTypeLinkService로 이름 수정 및 메서드 이름 수정

GunsanHaribo committed:
test: DeclaredTypeLinkService 테스트 코드 작성

- 양방향 매핑시 헬퍼메서드 필요합니다. DB에는 들어가는데 객체 그래프에는 반영이 안되는 경우가 있습니다.

GunsanHaribo committed:
refactor: 콜그래프 클래스들 이름 수정

GunsanHaribo committed:
refactor: MethodModifier, TypeKind 추출로직 추가

GunsanHaribo committed:
refactor: methodAccess, classAccess에서 추출한 정보를 enum으로 받도록 수정

GunsanHaribo committed:
test: PolymorphicEdgeComputer 테스트 작성

GunsanHaribo committed:
test: SourceCodeParser 테스트 수정

GunsanHaribo committed:
test: CodeMethodCallGraphService 테스트 수정 및 관련 파일이동

GunsanHaribo committed:
refactor: MethodPolymorphicDispatcher을 인터페이스연결, 상속 연결로 클래스 분리

GunsanHaribo committed:
test: MethodCallResolver 기초 테스트 하나 완료

GunsanHaribo committed:
refactor: ClassStructure 기준으로 분석 파이프라인 DTO 흐름 정리

- 파일 정보 DTO 과분할로 인한 혼란 제거
- rawText/path/start-endLine 포함한 ClassStructure로 결과 정규화
- 중간 단계에서 DTO 보정/끼어들기 로직 제거

GunsanHaribo committed:
refactor: DeclaredType의 DeclaredTypeImplementEdge의 외부노출 캡슐화

GunsanHaribo committed:
refactor: DeclaredType.external 생성시 레포를 통해서 생성하는 로직으로 변경

GunsanHaribo committed:
bug: 인터페이스 extends와 첫 구현체 탐색시 무한루프 버그 수정

GunsanHaribo committed:
bug: 상속 메서드 INHERITED_INTERNAL 연결시, 탐색 안되는 버그 수정 및 MethodOrigin 구체화

GunsanHaribo committed:
bug: 상속 메서드 INHERITED_INTERNAL 연결시 중복방문 수정

GunsanHaribo committed:
refactor: 엔티티 공통요소 shared 패키지로 리펙터링

GunsanHaribo committed:
bug: cascade 데이터 중복 저장 문제 임시조치

GunsanHaribo committed:
refactor: 엔티티 패키지구조 수정

생성책임을 내부로 옮기기위해서 정적팩터리 메서드를 일부 default 생성자로 변경 및 엔티티 패키지 구조 수정

GunsanHaribo committed:
refactor: ClassSuperDispatcher에서 메서드 생성책임을 Type으로 이전

GunsanHaribo committed:
refactor: InterfaceImplementationLinker에서 메서드 생성책임을 Type으로 이전

GunsanHaribo committed:
refactor: 클래스-메서드 이름 수정

GunsanHaribo committed:
refactor: MethodCallResolver 로직수정

GunsanHaribo committed:
refactor: 클래스 이동 및 주석으로 현재 문제점 기록

GunsanHaribo committed:
refactor: External 타입을 Content와 분리

- 레포에 Type을 저장하는 방향으로 진행

GunsanHaribo committed:
refactor: Content -> File로 이름 수정

GunsanHaribo committed:
refactor: 파일 위치 변경

GunsanHaribo committed:
refactor: 분석 단계를 4단계로 분리하고 resolve를 관계 연결 중심으로 단순화

기존에는 resolve 단계에서 심볼 생성, 외부 타입 생성,
계층 연결, 메서드 호출 연결을 함께 처리하고 있었음.

분석 과정을 다음 4단계로 재구성함:
1) Declared 심볼 저장 (Repo/Content/Type/Method)
2) External Type/Method 생성
3) Type 계층 관계 연결
4) Method Call 관계 연결

이를 통해 선언/생성과 관계 연결의 책임을 분리하고,
resolve 단계는 관계 연결을 중심으로 동작하도록 단순화함.
연결 과정에서 필요한 경우에만 최소한의 보정 생성 로직을 유지함.

GunsanHaribo committed:
refactor: 팩토리 클래스 이동

GunsanHaribo committed:
refactor: flyway 수정 및 깃레포 다운로드 & 타입구조 추출 실패시 디렉토리 삭제로직 추가

GunsanHaribo committed:
fix & refactor(method): Method 중복 생성 문제 해결 및 생성 로직 통일

Method 저장 이후 인메모리 인덱스(typeMethods)가 갱신되지 않아
동일 (type_id, method_signature)에 대해 중복 insert가 발생하는
문제를 수정하였다.

- MethodCallResolver에서 save 후 인덱스 미반영 문제 해결
- 상속 처리 및 인터페이스 구현 연결 로직에서 중복 생성 가능성 제거
- get → null → save → put 패턴을 computeIfAbsent 기반으로 통일

이를 통해
- UNIQUE(type_id, method_signature) 충돌 방지
- 인메모리 인덱스와 영속 상태 간 일관성 확보
- 그래프 구성 과정의 안정성 향상

하도록 구조를 개선하였다.

GunsanHaribo committed:
refactor: 테스트 통과하도록 수정

GunsanHaribo committed:
chore: 로그파일 생성 및 필요없는 테스트파일 삭제

GunsanHaribo committed:
refactor: 예외 발생시 @GlobalAdvice로 처리로 수정

GunsanHaribo committed:
test: TypeStructureResolver 예외 케이스 추가

GunsanHaribo committed:
Merge branch 'dev' into feat/14-class-useflow-analysis

# Conflicts:
#	api/src/main/java/com/opensourcereader/api/shared/config/HttpConfig.java
#	api/src/test/java/com/opensourcereader/api/facade/analysis/OpenSourceRepoFacadeTest.java
#	core/src/main/java/com/opensourcereader/core/analysis/entity/OpenSourceRepo.java
#	core/src/main/java/com/opensourcereader/core/analysis/repository/OpenSourceRepoRepository.java
#	core/src/main/java/com/opensourcereader/core/analysis/service/OpenSourceRepoService.java
#	core/src/main/java/com/opensourcereader/core/security/service/AuthService.java
#	core/src/main/java/com/opensourcereader/core/security/service/RefreshTokenService.java
#	core/src/main/java/com/opensourcereader/core/shared/config/HttpConfig.java
#	core/src/main/java/com/opensourcereader/core/shared/exception/OSRServerException.java
#	core/src/main/java/com/opensourcereader/core/user/entity/User.java
#	core/src/main/java/com/opensourcereader/core/user/service/UserServiceImpl.java
#	core/src/test/java/com/opensourcereader/core/security/service/AuthServiceUnitTest.java
#	core/src/test/java/com/opensourcereader/core/security/service/RefreshTokenServiceUnitTest.java

GunsanHaribo committed:
refactor: 멀티모듈 프로젝트 core(core-api, core-domain)으로 변경

GunsanHaribo committed:
refactor: db를 postgres로 변경 및 바이너리 파일은 DB에 저장하지 않는걸로 수정

GunsanHaribo committed:
refactor: 패키지 이름 변경 및 패키지 이동

GunsanHaribo committed:
refactor: 테스트 설정 파일도 postgres로 변경

GunsanHaribo committed:
bug: 멀티 모듈 설정 수정
